packages/client/src/util/toDataValue.ts

```diff
@@ -25,25 +31,48 @@ import { isWireObjectSet } from "./WireObjectSet.js";
  *
  * @see DataValue for the expected payloads
  */
-export function toDataValue(value: any): DataValue {
+export async function toDataValue(
+  value: any,
+  client: MinimalClient,
+): Promise<DataValue> {
   if (value == null) {
     // typeof null is 'object' so do this first
     return value;
   }
 
   // arrays and sets are both sent over the wire as arrays
   if (Array.isArray(value) || value instanceof Set) {
-    return Array.from(value, toDataValue);
+    const promiseArray = Array.from(
+      value,
+      async (innerValue) => await toDataValue(innerValue, client),
+    );
+    return Promise.all(promiseArray);
   }
 
   // attachments just send the rid directly
   if (isAttachment(value)) {
     return value.rid;
   }
 
+  // For uploads, we need to upload ourselves first to get the RID of the attachment
+  if (isAttachmentUpload(value)) {
+    const attachment = await Ontologies.Attachments.uploadAttachment(
```

Comment:
This seemed like the best place to put this to follow the existing pattern. Only downside is had to modify everything else in this to be async, and nothing else really needs the client context.